sklearn/tree/_tree.pyx

```diff
@@ -3338,6 +3338,158 @@ cdef class Tree:
 
         return out
 
+    cpdef np.ndarray apply_depth(self, object X):
+        """Finds the depth level the of the terminal region (=leaf node) for each sample in X."""
+        if issparse(X):
+            return self._apply_depth_sparse_csr(X)
+        else:
+            return self._apply_depth_dense(X)
+
+    cdef inline np.ndarray _apply_depth_dense(self, object X):
+        """Finds the depth level of the terminal region (=leaf node) for each sample in X."""
+
+        # Check input
+        if not isinstance(X, np.ndarray):
+            raise ValueError("X should be in np.ndarray format, got %s"
+                             % type(X))
+
+        if X.dtype != DTYPE:
+            raise ValueError("X.dtype should be np.float32, got %s" % X.dtype)
+
+        # Extract input
+        cdef np.ndarray X_ndarray = X
+        cdef DTYPE_t* X_ptr = <DTYPE_t*> X_ndarray.data
+        cdef SIZE_t X_sample_stride = <SIZE_t> X.strides[0] / <SIZE_t> X.itemsize
+        cdef SIZE_t X_fx_stride = <SIZE_t> X.strides[1] / <SIZE_t> X.itemsize
+        cdef SIZE_t n_samples = X.shape[0]
+
+        # Initialize output
+        cdef np.ndarray[DOUBLE_t] depth = np.zeros((n_samples,), dtype=np.float64)
+        cdef DOUBLE_t* depth_ptr = <DOUBLE_t*> depth.data
+
+
+        # Initialize auxiliary data-structure
+        cdef Node* node = NULL
+        cdef SIZE_t i = 0
+        cdef DOUBLE_t cpt = 0.
+        cdef DOUBLE_t harmonic_number = 0.
+        cdef SIZE_t n = 0
+
+        with nogil:
+            for i in range(n_samples):
+                node = self.nodes
+                cpt = 0.
+                # While node not a leaf
+                while node.left_child != _TREE_LEAF:
+                    # ... and node.right_child != _TREE_LEAF:
+                    if X_ptr[X_sample_stride * i +
+                             X_fx_stride * node.feature] <= node.threshold:
+                        node = &self.nodes[node.left_child]
+                    else:
+                        node = &self.nodes[node.right_child]
+                    cpt += 1.
+                n = <SIZE_t> node.n_node_samples  # Number of samples left in the leaf
+                if n > 1:
+                    # Add the average path length in a n samples iTree, which is equal to
+                    # the average path length of an unsuccessful BST search - since the
+                    # latter has the same structure as an isolation tree.
+                    harmonic_number = <DOUBLE_t> ln(n) 
+                    harmonic_number += 0.5772156649
+                    depth_ptr[i] = <DOUBLE_t> cpt + 2. * harmonic_number - 2. * (n - 1.) / n
```

Comment:
So in case we have several samples in the leaf, the depth of sample `i` is `cpt` + the average path length, is that right? If yes, then this should be mentioned in the docstring as well. 

Also, could you factor out this piece of code, since it is common to both dense and sparse cases?